### ISSUE TITLE
Trying to close a never connected websocket returns an endless future

### ISSUE THREAD

**Issue by Bungeefan** [dart-lang/http#1465]
### Use case:
We tried to listen to a web socket stream after we created it.
The web socket can't reach the server and fails with an error in the `onError` handler.

Finally, when trying to close such a web socket with `await webSocket?.sink.close();`, it hangs in an endless future.

### Research:
As far my research went, I found the `done` getter in the `_CompleterSink` to be the underlying problem.

The following code returns our endless future:
https://github.com/dart-lang/web_socket_channel/blob/4b46c0c4196a5e76c2b0e2589ed37de247d35938/lib/src/sink_completer.dart#L75-L83

It returns a future which waits for the destination sink, which only gets completed when `_setDestinationSink` is called.
Which in our case never gets called, because the destination sink was never successfully created, therefore the future never completes.
Call that would complete the future, but only after successful creation: https://github.com/dart-lang/web_socket_channel/blob/4b46c0c4196a5e76c2b0e2589ed37de247d35938/lib/io.dart#L73

### Possible solution (First idea):
I successfully tested the following code example which seems to fix it, however I am not sure if there are any possible side effects with this "solution".

This checks in the close method if there is any current destination sink, if there isn't the caller probably isn't interested in waiting for one to complete. Otherwise, behave as before.

```dart
@override
Future close([int? closeCode, String? closeReason]) {
  if (_canSendDirectly) {
    _destinationSink!.close(closeCode, closeReason);
  } else {
    _closeCode = closeCode;
    _closeReason = closeReason;
    _ensureController().close();
  }
  if (_destinationSink == null) {
    return Future.value();
  }
  return done;
}
```
This would replace the following current code:
https://github.com/dart-lang/web_socket_channel/blob/4b46c0c4196a5e76c2b0e2589ed37de247d35938/lib/src/sink_completer.dart#L111-L121

### Additional notes:

I also found another issue (https://github.com/dart-lang/web_socket_channel/issues/185) which seems to run in the same or a similar issue and the only comment was to not await the future, which isn't really a solution.
